### ISSUE TITLE
Numpy interface can fail with non-writeable arrays.

### ISSUE THREAD

**Issue by mforbes** [pyFFTW/pyFFTW#92]
The following code fails if it is called before a plan is cached:

```python
import numpy as np
from pyfftw.interfaces.numpy_fft import fat
y = np.zeros(4) + 1j  # Needs to be complex for failure
y.flags.writeable = False
fft(y)
```

```
In [5]: fft(y)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
...
.../pyfftw/interfaces/_utils.pyc in _Xfftn(a, s, axes, overwrite_input, planner_effort, threads, auto_align_input, auto_contiguous, calling_func)
     81         a_copy = a.copy()
     82 
---> 83         FFTW_object = getattr(builders, calling_func)(*args)
     84 
     85         # Only copy if the input array is what was actually used
...
.../pyfftw/builders/_utils.pyc in _Xfftn(a, s, axes, overwrite_input, planner_effort, threads, auto_align_input, auto_contiguous, avoid_copy, inverse, real)
    171         if not avoid_copy:
    172             # Copy the data back into the (likely) destroyed array
--> 173             FFTW_object.get_input_array()[:] = a_copy
    174 
    175     return FFTW_object

ValueError: assignment destination is read-only
```

The problem is that the plan is made using the original array, which in this case is not writeable.  Since a copy is made, why not use it for planning instead since the copy will always be writeable?

The minimal change would be to set and restore the writeable flag:

```patch
82a83,85
>         # Input array must be writeable...
>         writeable, a.flags.writeable = a.flags.writeable, True
> 
89a93,95
>         # Restore writeable flag
>         a.flags.writeable = writeable
> 
```

Using the copy instead seems to work but there may be subtleties I am not aware of.

```patch
24c24
< 
---
> import numpy as np
81c81
<         a_copy = a.copy()
---
>         args = (np.empty_like(a),) + args[1:]
85,89d84
<         # Only copy if the input array is what was actually used
<         # (otherwise it shouldn't be overwritten)
<         if FFTW_object.get_input_array() is a:
<             a[:] = a_copy
< 
93c88
<         output_array = FFTW_object()
---
>         output_array = FFTW_object(input_array=a) 
```

**Comment by hgomersall** [comment]
This is an interesting one. I think the only safe behaviour is to not touch an array that is marked read only, in which case, we should basically require, at least for the numpy interface, that the input have its writeable flag as True. This is definitely a bug, so thanks.

This is not the case for the direct FFTW object, for which we can use FFTW's arguments for not touching the array.

**Comment by mforbes** [comment]
It is my impression that the input is modified only during the planning phase, which can be remedied by using a copy to do the planning.  I regularly lock my input arrays to prevent unintentional mutation in my code, and it would be quite inconvenient to have to special-case using the numpy interface each time.  Are there subtleties with the planning that require using the actual input array rather than a copy?

**Comment by hgomersall** [comment]
It might be disastrous to start writing say a memory mapped array which has its writable flag set to false. IMO it's an egregious breach of contract to ignore flags like that.

Special casing is easy with a short wrapper function if that's desired, but it should not be up to pyfftw to make assumptions about what is meant by `writable =False`.

**Comment by hgomersall** [comment]
Oh I missed the core of your point. Yes, copying to plan is not a bad idea. I'll think some more about it.

Certain functions always destroy their input, so would require a copy every time.

**Comment by mforbes** [comment]
The only thing I was thinking might be an issue with a copy is if the input array is strided.  Copies will be contiguous but using `empty_like()` preserves C vs F ordering for example.

**Comment by hgomersall** [comment]
I think this is non-trivial to fix. It would be a bit of a hack and IMO is much better to be handled from the calling end with the correct policy set there (either copy, or enable `writable=True` or something else). The problem being is `writeable=True` is IFAICT not a common flag to set and as such would require a disproportionate amount of corner case hackery without necessarily touching on all the reasons why it's done.

It is not particularly onerous to create a wrapper function to do the copying etc, so I think the correct thing to do from pyfftw's end is to raise on arrays with `writable=False`.

OTOH, if you can show how this can work without lots of additional code (and ideally a PR) I'm happy to listen,

**Comment by mforbes** [comment]
I have posted a suggestion as PR #93.  This seems fairly trivial and `python setup.py test` passes, but I am not sure I understand all the subtle points.

**Comment by hgomersall** [comment]
Fixed in #93 .
